File: graph_shortest_path/dijkstra/src/DijkstraDemo.rb
# Ruby
# グラフの最短経路: ダイクストラ法 (dijkstra)

class GraphData
  def initialize
    # 隣接ノードとその辺の重みを格納します。
    # キーは頂点、値はその頂点に隣接する頂点と重みの配列です。
    # 例: { 'A' => [['B', 1], ['C', 4]], 'B' => [['A', 1], ['C', 2], ['D', 5]], ... }
    @data = {}
  end

  def get
    # グラフの内部データを取得します。
    @data
  end

  def get_vertices
    # グラフの全頂点をリストとして返します。
    @data.keys
  end

  def get_edges
    # グラフの全辺をリストとして返します。
    # 無向グラフの場合、[u, v, weight] の形式で返します。
    # 重複を避けるためにセットを使用します。
    edges = Set.new
    @data.each do |vertex, neighbors|
      neighbors.each do |neighbor, weight|
        # 辺を正規化してセットに追加 (小さい方の頂点を最初にするなど)
        edge = [vertex, neighbor].sort
        edges.add([edge[0], edge[1], weight]) # [u, v, weight] の形式で格納
      end
    end
    edges.to_a
  end

  def get_neighbors(vertex)
    # 指定された頂点の隣接ノードと辺の重みのリストを返します。
    # 形式: [[隣接頂点, 重み], ...]
    if @data.key?(vertex)
      @data[vertex]
    else
      nil # 頂点が存在しない場合はnilを返す
    end
  end

  def add_vertex(vertex)
    # 新しい頂点をグラフに追加します。
    unless @data.key?(vertex)
      @data[vertex] = []
      return true
    end
    # 既に存在する場合は追加しないがtrueを返す（変更なしでも成功とみなす）
    true
  end

  def add_edge(vertex1, vertex2, weight)
    # 両頂点間に辺を追加します。重みを指定します。
    # 頂点がグラフに存在しない場合は追加します。
    add_vertex(vertex1) unless @data.key?(vertex1)
    add_vertex(vertex2) unless @data.key?(vertex2)
    
    # 両方向に辺を追加する（無向グラフ）
    # 既に同じ辺が存在する場合は重みを更新するかどうか？
    # 今回は単純に追加する（重複辺はプリム法では問題にならないが、データとしては綺麗でないかも）
    # ここでは、同じ頂点間の辺が存在しない場合のみ追加するように修正します。
    
    # vertex1 -> vertex2 の辺を追加（重み付き）
    edge_exists_v1v2 = false
    @data[vertex1].each_with_index do |(neighbor, _), i|
      if neighbor == vertex2
        @data[vertex1][i] = [vertex2, weight] # 既に存在する場合は重みを更新
        edge_exists_v1v2 = true
        break
      end
    end
    @data[vertex1] << [vertex2, weight] unless edge_exists_v1v2

    # vertex2 -> vertex1 の辺を追加（重み付き）
    edge_exists_v2v1 = false
    @data[vertex2].each_with_index do |(neighbor, _), i|
      if neighbor == vertex1
        @data[vertex2][i] = [vertex1, weight] # 既に存在する場合は重みを更新
        edge_exists_v2v1 = true
        break
      end
    end
    @data[vertex2] << [vertex1, weight] unless edge_exists_v2v1
    
    true
  end

  def clear
    # グラフを空にします。
    @data = {}
    true
  end

  def get_shortest_path(start_vertex, end_vertex, heuristic)
    unless @data.key?(start_vertex) && @data.key?(end_vertex)
      puts "ERROR: 開始頂点 '#{start_vertex}' または 終了頂点 '#{end_vertex}' がグラフに存在しません。"
      return [nil, Float::INFINITY]
    end

    # 距離を初期化: 全ての頂点への距離を無限大に設定し、開始頂点のみ0とする
    distances = {}
    @data.each_key { |vertex| distances[vertex] = Float::INFINITY }
    distances[start_vertex] = 0

    # 経路を記録するための辞書: 各頂点に至る直前の頂点を保持
    predecessors = {}
    @data.each_key { |vertex| predecessors[vertex] = nil }

    # 優先度付きキュー: [距離, 頂点] の配列を格納し、距離が小さい順に取り出す
    # Rubyの場合、PriorityQueueが標準ライブラリにはないため、配列を使って実装します
    priority_queue = [[0, start_vertex]] # [開始頂点への距離, 開始頂点]

    until priority_queue.empty?
      # 優先度付きキューから最も距離の小さい頂点を取り出す
      priority_queue.sort! # 距離順にソート
      current_distance, current_vertex = priority_queue.shift # 最小距離の要素を取り出す

      # 取り出した頂点への距離が、すでに記録されている距離より大きい場合は、
      # より短い経路が既に見つかっているためスキップ
      next if current_distance > distances[current_vertex]

      # 終了頂点に到達したら探索終了
      break if current_vertex == end_vertex # 最短経路が見つかった

      # 現在の頂点から到達可能な隣接頂点を探索
      # self.get_neighbors(current_vertex) は [[neighbor, weight], ...] のリストを返す
      get_neighbors(current_vertex).each do |neighbor, weight|
        distance_through_current = distances[current_vertex] + weight

        # より短い経路が見つかった場合
        if distance_through_current < distances[neighbor]
          distances[neighbor] = distance_through_current
          predecessors[neighbor] = current_vertex
          # 優先度付きキューに隣接頂点を追加または更新
          # ダイクストラ法では heuristic は使用しない (または h=0)
          priority_queue.push([distance_through_current, neighbor])
        end
      end
    end

    # 終了頂点への最短距離が無限大のままなら、到達不可能
    if distances[end_vertex] == Float::INFINITY
      puts "INFO: 開始頂点 '#{start_vertex}' から 終了頂点 '#{end_vertex}' への経路は存在しません。"
      return [nil, Float::INFINITY]
    end

    # 最短経路を再構築
    path = []
    current = end_vertex
    while current
      path.unshift(current) # 経路の先頭に追加（逆順を避けるため）
      current = predecessors[current]
    end

    # 開始ノードから開始されていることを確認
    if path[0] != start_vertex
      # これが発生するのは、開始頂点が存在しないか、到達不能な場合だが、
      # 前段のチェックで大部分はカバーされているはず。
      # ここに来る場合は、特殊なケース（例えば孤立した開始点と終了点）が考えられる。
      return [nil, Float::INFINITY]
    end

    [path, distances[end_vertex]]
  end
end

# ヒューリスティック関数 (この例では常に0、ダイクストラ法と同じ)
# 実際のA*では、問題に応じた適切な推定関数を使用する必要があります。
def dummy_heuristic(u, v)
  # u と v の間に何らかの推定距離を計算する関数
  # ここではダミーとして常に0を返す
  0
end

def main
  puts "Dijkstra -----> start"

  require 'set'
  graph_data = GraphData.new

  graph_data.clear
  input_list = [['A', 'B', 4], ['B', 'C', 3], ['B', 'D', 2], ['D', 'A', 1], ['A', 'C', 2], ['B', 'D', 2]]
  input_list.each do |input|
    graph_data.add_edge(*input)
  end
  puts "\nグラフの頂点: #{graph_data.get_vertices}"
  puts "グラフの辺 (重み付き): #{graph_data.get_edges}"
  input = ['A', 'B']
  shortest_path = graph_data.get_shortest_path(input[0], input[1], method(:dummy_heuristic))
  puts "経路#{input[0]}-#{input[1]} の最短経路は #{shortest_path[0]} (重み: #{shortest_path[1]})"

  graph_data.clear
  input_list = [['A', 'B', 4], ['C', 'D', 4], ['E', 'F', 1], ['F', 'G', 1]]
  input_list.each do |input|
    graph_data.add_edge(*input)
  end
  puts "\nグラフの頂点: #{graph_data.get_vertices}"
  puts "グラフの辺 (重み付き): #{graph_data.get_edges}"
  input = ['A', 'B']
  shortest_path = graph_data.get_shortest_path(input[0], input[1], method(:dummy_heuristic))
  puts "経路#{input[0]}-#{input[1]} の最短経路は #{shortest_path[0]} (重み: #{shortest_path[1]})"

  graph_data.clear
  input_list = [['A', 'B', 4], ['B', 'C', 3], ['D', 'E', 5]]
  input_list.each do |input|
    graph_data.add_edge(*input)
  end
  puts "\nグラフの頂点: #{graph_data.get_vertices}"
  puts "グラフの辺 (重み付き): #{graph_data.get_edges}"
  input = ['A', 'D']
  shortest_path = graph_data.get_shortest_path(input[0], input[1], method(:dummy_heuristic))
  puts "経路#{input[0]}-#{input[1]} の最短経路は #{shortest_path[0]} (重み: #{shortest_path[1]})"

  graph_data.clear
  input_list = []
  input_list.each do |input|
    graph_data.add_edge(*input)
  end
  puts "\nグラフの頂点: #{graph_data.get_vertices}"
  puts "グラフの辺 (重み付き): #{graph_data.get_edges}"
  input = ['A', 'B']
  shortest_path = graph_data.get_shortest_path(input[0], input[1], method(:dummy_heuristic))
  puts "経路#{input[0]}-#{input[1]} の最短経路は #{shortest_path[0]} (重み: #{shortest_path[1]})"

  puts "\nDijkstra <----- end"
end

if __FILE__ == $PROGRAM_NAME
  main
end

File: graph_shortest_path/dijkstra/src/DijkstraDemo.go
// Go
// グラフの最短経路: ダイクストラ法 (dijkstra)

package main

import (
	"container/heap"
	"fmt"
)

// GraphData はグラフのデータ構造を表します
type GraphData struct {
	data map[string][]Edge
}

// Edge はグラフの辺を表します
type Edge struct {
	Neighbor string
	Weight   float64
}

// NewGraphData は新しいGraphDataを作成します
func NewGraphData() *GraphData {
	return &GraphData{
		data: make(map[string][]Edge),
	}
}

// Get はグラフの内部データを取得します
func (g *GraphData) Get() map[string][]Edge {
	return g.data
}

// GetVertices はグラフの全頂点をリストとして返します
func (g *GraphData) GetVertices() []string {
	vertices := make([]string, 0, len(g.data))
	for vertex := range g.data {
		vertices = append(vertices, vertex)
	}
	return vertices
}

// GetEdges はグラフの全辺をリストとして返します
func (g *GraphData) GetEdges() []struct {
	Vertex1, Vertex2 string
	Weight           float64
} {
	edges := make(map[string]struct {
		Vertex1, Vertex2 string
		Weight           float64
	})

	for vertex, neighbors := range g.data {
		for _, edge := range neighbors {
			// 辺を正規化（小さい方の頂点を最初にする）
			v1, v2 := vertex, edge.Neighbor
			if v1 > v2 {
				v1, v2 = v2, v1
			}
			key := v1 + "-" + v2
			edges[key] = struct {
				Vertex1, Vertex2 string
				Weight           float64
			}{v1, v2, edge.Weight}
		}
	}

	result := make([]struct {
		Vertex1, Vertex2 string
		Weight           float64
	}, 0, len(edges))

	for _, edge := range edges {
		result = append(result, edge)
	}

	return result
}

// GetNeighbors は指定された頂点の隣接ノードと辺の重みのリストを返します
func (g *GraphData) GetNeighbors(vertex string) []Edge {
	if neighbors, exists := g.data[vertex]; exists {
		return neighbors
	}
	return nil
}

// AddVertex は新しい頂点をグラフに追加します
func (g *GraphData) AddVertex(vertex string) bool {
	if _, exists := g.data[vertex]; !exists {
		g.data[vertex] = []Edge{}
	}
	return true
}

// AddEdge は両頂点間に辺を追加します
func (g *GraphData) AddEdge(vertex1, vertex2 string, weight float64) bool {
	if _, exists := g.data[vertex1]; !exists {
		g.AddVertex(vertex1)
	}
	if _, exists := g.data[vertex2]; !exists {
		g.AddVertex(vertex2)
	}

	// vertex1 -> vertex2 の辺を追加
	edgeExistsV1V2 := false
	for i, edge := range g.data[vertex1] {
		if edge.Neighbor == vertex2 {
			g.data[vertex1][i].Weight = weight // 既に存在する場合は重みを更新
			edgeExistsV1V2 = true
			break
		}
	}
	if !edgeExistsV1V2 {
		g.data[vertex1] = append(g.data[vertex1], Edge{vertex2, weight})
	}

	// vertex2 -> vertex1 の辺を追加
	edgeExistsV2V1 := false
	for i, edge := range g.data[vertex2] {
		if edge.Neighbor == vertex1 {
			g.data[vertex2][i].Weight = weight // 既に存在する場合は重みを更新
			edgeExistsV2V1 = true
			break
		}
	}
	if !edgeExistsV2V1 {
		g.data[vertex2] = append(g.data[vertex2], Edge{vertex1, weight})
	}

	return true
}

// Clear はグラフを空にします
func (g *GraphData) Clear() bool {
	g.data = make(map[string][]Edge)
	return true
}

// PriorityQueueItem は優先度付きキューのアイテムを表します
type PriorityQueueItem struct {
	vertex   string
	distance float64
	index    int
}

// PriorityQueue は優先度付きキューを実装します
type PriorityQueue []*PriorityQueueItem

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].distance < pq[j].distance
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*PriorityQueueItem)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil  // メモリリークを避けるため
	item.index = -1 // 安全のため
	*pq = old[0 : n-1]
	return item
}

// 無限大を表す定数
const Infinity = float64(int(^uint(0) >> 1))

// GetShortestPath は開始頂点から終了頂点までの最短経路を返します
func (g *GraphData) GetShortestPath(startVertex, endVertex string, heuristic func(string, string) float64) ([]string, float64) {
	if _, exists := g.data[startVertex]; !exists {
		fmt.Printf("ERROR: 開始頂点 '%s' がグラフに存在しません。\n", startVertex)
		return nil, Infinity
	}
	if _, exists := g.data[endVertex]; !exists {
		fmt.Printf("ERROR: 終了頂点 '%s' がグラフに存在しません。\n", endVertex)
		return nil, Infinity
	}

	// 距離を初期化: 全ての頂点への距離を無限大に設定し、開始頂点のみ0とする
	distances := make(map[string]float64)
	for vertex := range g.data {
		distances[vertex] = Infinity
	}
	distances[startVertex] = 0

	// 経路を記録するための辞書: 各頂点に至る直前の頂点を保持
	predecessors := make(map[string]string)
	for vertex := range g.data {
		predecessors[vertex] = ""
	}

	// 優先度付きキューを初期化
	pq := make(PriorityQueue, 0)
	heap.Init(&pq)
	heap.Push(&pq, &PriorityQueueItem{
		vertex:   startVertex,
		distance: 0,
	})

	for pq.Len() > 0 {
		// 優先度付きキューから最も距離の小さい頂点を取り出す
		item := heap.Pop(&pq).(*PriorityQueueItem)
		currentVertex := item.vertex
		currentDistance := item.distance

		// 取り出した頂点への距離が、すでに記録されている距離より大きい場合は、
		// より短い経路が既に見つかっているためスキップ
		if currentDistance > distances[currentVertex] {
			continue
		}

		// 終了頂点に到達したら探索終了
		if currentVertex == endVertex {
			break // 最短経路が見つかった
		}

		// 現在の頂点から到達可能な隣接頂点を探索
		for _, edge := range g.GetNeighbors(currentVertex) {
			distanceThroughCurrent := distances[currentVertex] + edge.Weight

			// より短い経路が見つかった場合
			if distanceThroughCurrent < distances[edge.Neighbor] {
				distances[edge.Neighbor] = distanceThroughCurrent
				predecessors[edge.Neighbor] = currentVertex
				// 優先度付きキューに隣接頂点を追加または更新
				heap.Push(&pq, &PriorityQueueItem{
					vertex:   edge.Neighbor,
					distance: distanceThroughCurrent,
				})
			}
		}
	}

	// 終了頂点への最短距離が無限大のままなら、到達不可能
	if distances[endVertex] == Infinity {
		fmt.Printf("INFO: 開始頂点 '%s' から 終了頂点 '%s' への経路は存在しません。\n", startVertex, endVertex)
		return nil, Infinity
	}

	// 最短経路を再構築
	path := make([]string, 0)
	current := endVertex
	for current != "" {
		path = append([]string{current}, path...) // 先頭に追加（逆順にする）
		current = predecessors[current]
	}

	// 開始ノードから開始されていることを確認
	if len(path) == 0 || path[0] != startVertex {
		return nil, Infinity
	}

	return path, distances[endVertex]
}

// ヒューリスティック関数 (常に0を返す、ダイクストラ法と同じ)
func dummyHeuristic(u, v string) float64 {
	return 0
}

func main() {
	fmt.Println("Dijkstra -----> start")

	graphData := NewGraphData()

	graphData.Clear()
	inputList := []struct {
		v1, v2 string
		weight float64
	}{
		{"A", "B", 4},
		{"B", "C", 3},
		{"B", "D", 2},
		{"D", "A", 1},
		{"A", "C", 2},
		{"B", "D", 2},
	}
	for _, input := range inputList {
		graphData.AddEdge(input.v1, input.v2, input.weight)
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	input := [2]string{"A", "B"}
	shortestPath, weight := graphData.GetShortestPath(input[0], input[1], dummyHeuristic)
	fmt.Printf("経路%s-%s の最短経路は %v (重み: %v)\n", input[0], input[1], shortestPath, weight)

	graphData.Clear()
	inputList = []struct {
		v1, v2 string
		weight float64
	}{
		{"A", "B", 4},
		{"C", "D", 4},
		{"E", "F", 1},
		{"F", "G", 1},
	}
	for _, input := range inputList {
		graphData.AddEdge(input.v1, input.v2, input.weight)
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	input = [2]string{"A", "B"}
	shortestPath, weight = graphData.GetShortestPath(input[0], input[1], dummyHeuristic)
	fmt.Printf("経路%s-%s の最短経路は %v (重み: %v)\n", input[0], input[1], shortestPath, weight)

	graphData.Clear()
	inputList = []struct {
		v1, v2 string
		weight float64
	}{
		{"A", "B", 4},
		{"B", "C", 3},
		{"D", "E", 5},
	}
	for _, input := range inputList {
		graphData.AddEdge(input.v1, input.v2, input.weight)
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	input = [2]string{"A", "D"}
	shortestPath, weight = graphData.GetShortestPath(input[0], input[1], dummyHeuristic)
	fmt.Printf("経路%s-%s の最短経路は %v (重み: %v)\n", input[0], input[1], shortestPath, weight)

	graphData.Clear()
	inputList = []struct {
		v1, v2 string
		weight float64
	}{}
	for _, input := range inputList {
		graphData.AddEdge(input.v1, input.v2, input.weight)
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	input = [2]string{"A", "B"}
	shortestPath, weight = graphData.GetShortestPath(input[0], input[1], dummyHeuristic)
	fmt.Printf("経路%s-%s の最短経路は %v (重み: %v)\n", input[0], input[1], shortestPath, weight)

	fmt.Println("\nDijkstra <----- end")
}